dozer/cogs/filter.py

```diff
@@ -0,0 +1,259 @@
+            return
+        new_filter = WordFilter(guild_id=ctx.guild.id, pattern=pattern, friendly_name=friendly_name or pattern)
+        with db.Session() as session:
+            session.add(new_filter)
+        embed = discord.Embed()
+        embed.title = "Filter {} added".format(new_filter.id)
+        embed.description = "A new filter with the name `{}` was added.".format(friendly_name or pattern)
+        embed.add_field(name="Pattern", value="`{}`".format(pattern))
+        await ctx.send(embed=embed)
+        self.load_filters(ctx.guild.id)
+
+    add.example_usage = "`{prefix}filter whitelist add Swear` - Makes it so that \"Swear\" will be filtered"
+
+    @guild_only()
+    @has_permissions(manage_guild=True)
+    @filter.command()
+    async def remove(self, ctx, id):
+        """Remove a pattern from the filter list."""
+        with db.Session() as session:
+            result = session.query(WordFilter).filter_by(id = id).one_or_none()
+            if result is None:
+                await ctx.send("Filter ID {} not found!".format(id))
+                return
+            if result.guild_id != ctx.guild.id:
+                await ctx.send("That Filter does not belong to this guild.")
+                return
+            result.enabled = False
+            await ctx.send("Filter `{}` with name `{}` deleted.".format(result.id, result.friendly_name))
+            self.load_filters(ctx.guild.id)
+
+    remove.example_usage = "`{prefix}filter remove 7` - Disables filter with ID 7"
+
+    @guild_only()
+    @has_permissions(manage_guild=True)
+    @filter.command(name="dm")
+    async def dm_config(self, ctx, config: bool):
+        """Set whether filter words should be DMed when used in bot messages"""
+        with db.Session() as session:
+            result = session.query(WordFilterSetting).filter_by(guild_id=ctx.guild.id, setting_type="dm") \
+                .one_or_none()
+            try:
+                before_setting = result.value
+                result.value = config
+            except AttributeError:
+                before_setting = None
+                result = WordFilterSetting(guild_id=ctx.guild.id, setting_type="dm", value=config)
+                session.add(result)
+            await ctx.send(
+                "The DM setting for this guild has been changed from {} to {}.".format(before_setting, result.value))
+
+    dm_config.example_usage = "`{prefix}filter dm_config True` - Makes all messages containining filter lists to be sent through DMs"
+
+    @guild_only()
+    @filter.group(invoke_without_command=True)
+    async def whitelist(self, ctx):
+        """List all whitelisted roles for this server"""
+        with db.Session() as session:
+            results = session.query(WordFilterRoleWhitelist).filter_by(guild_id=ctx.guild.id).all()
+            role_objects = (discord.utils.get(ctx.guild.roles, id=db_role.role_id) for db_role in results)
```

Comment:
This iterates over the guild's entire role list once for every whitelisted role. I would set this up in a large-side-of-reasonable server (say, 150 roles), whitelist 5-10 of them, and see how long this takes.

Reply:
Added 150 roles, whitelisted 10 of them, executed virtually instantly.

Reply:
Since #135, finding a role by ID is now O(1) with `Guild.get_role`. It can be virtually instant-er.